### PR TITLE
Update rotatedRectangleIntersection function to calculate near to origin

### PR DESCRIPTION
**Pull request by gasparitiago**
This commit changes the rotatedRectangleIntersection function in order to calculate the intersection of two rectangles considering that they are shifted near the coordinates origin (0, 0).

This commit solves the problem in some assertions from rotatedRectangleIntersection when dealing with rectangles far from origin.

Original issue: https://github.com/opencv/opencv/issues/19824

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=linux,docs
```

### CODE REVIEW

**Comment by crackwitz** [comment]
does this actually make a difference? it only applies double precision to the calculation itself but the result is still just FP32.

the demo code uses values like 4002326. there is simply no precision in the single single digits or right of the decimal point.

**Comment by gasparitiago** [comment]
@crackwitz 
> does this actually make a difference?

Yes. When testing the code provided by @jbwintergest here: https://github.com/opencv/opencv/issues/19824:

```
import cv2 as cv
import numpy as np


def box(box):
    return ((box[0], box[1]), (box[2], box[3]), box[4])


box_1 = np.array([ 246805.033  , 4002326.94   ,      26.40587,       6.20026,     -62.10156])
box_2 = np.array([ 246805.122  , 4002326.59   ,      27.4821 ,       8.5361 ,     -56.33761])

try:
    print(cv.rotatedRectangleIntersection(box(box_1), box(box_2)))  # Assertion error
except Exception as e:
    print(e)

offset = (246805, 4002326, 0, 0, 0)

box_1 -= offset
box_2 -= offset

print(cv.rotatedRectangleIntersection(box(box_1), box(box_2)))  # that works
```

when the function `rotatedRectangleIntersection`, creates the `pts1` and `pts2` arrays and populate them, like this:

```
    Point2f pts1[4], pts2[4];
    ...
    rect1.points(pts1);
    rect2.points(pts2);
``` 

This is the content of `rect1`, `rect2`, `pts1[0]` and `pts2[0]` before and after this commit:

**Before:**
```
rect1.center.x
246805.03125
rect1.center.y
4002327

rect1.size.width
26.405870437622070312
rect1.size.height
6.200260162353515625

rect1.angle
-62.101558685302734375

rect2.center.x
246805.125
rect2.center.y
4002326.5

rect2.size.width
27.482099533081054688
rect2.size.height
8.5361003875732421875

rect2.angle
-56.33760833740234375

pts1[0].x pts1[0].y
246801.59375 4002340

pts2[0].x pts2[0].y
246801.0625 4002340.25
```

**After:**
```
rect1.center.x
246805.03125
rect1.center.y
4002327

rect1.size.width
26.405870437622070312
rect1.size.height
6.200260162353515625

rect1.angle
-62.101558685302734375

rect2.center.x
246805.125
rect2.center.y
4002326.5

rect2.size.width
27.482099533081054688
rect2.size.height
8.5361003875732421875

rect2.angle
-56.33760833740234375

pts1[0].x pts1[0].y
246801.59375 4002340.25

pts2[0].x pts2[0].y
246801.0625 4002340.25
``` 

When the function `rotatedRectangleIntersection` code considers if two points are the same or not, points that are too far away from the origin (high x or y) are considered overlapped in the **before** code.

**Comment by crackwitz** [comment]
the only difference between your two texts is that pts1[0].y is completely integer in the first and has a .25 decimal part in the second:

image

the algorithm itself will likely need fixes, not just the `Rect::points()` method

**Comment by asenyaev** [comment]
jenkins cn please retry a build

**Comment by asenyaev** [comment]
jenkins cn please retry a build

**Comment by vpisarev** [comment]
@gasparitiago, thank you for the contribution! I wonder if you could also modify `rotatedRectangleIntersection` function to shift intersected rectangles closer to the coordinate origin (and then shift the intersection back)? This way we will get much more robust algorithm

**Comment by gasparitiago** [comment]
Hello @vpisarev, I'm doing it. When I finish I will update this PR and let you know.

To be honest I will need some help on how to test the solution.

**Comment by gasparitiago** [comment]
@vpisarev can you take a look at my last commit?

File: modules/imgproc/src/intersection.cpp [hunk 1]
```diff
@@ -51,10 +51,20 @@ int rotatedRectangleIntersection( const RotatedRect& rect1, const RotatedRect& r
 {
     CV_INSTRUMENT_REGION();
```
Comment (alalek) [inline modules/imgproc/src/intersection.cpp:52]: There are several return statements from this code.
Some of them are **NOT** handled properly (e.g., line 97/107).

Robust way is to to create wrapper which would call the original code:
- make existed implementation as internal:
  ```
  static
  int rotatedRectangleIntersection_(
          const RotatedRect& rect1,
          const RotatedRect& rect2,
          CV_OUT std::vector <Point2f>& intersection)
  {
      ... almost the same existed code ...
  }
  ```
- add wrapper below which shifts input and adjusts results:
  ```
  int rotatedRectangleIntersection(const RotatedRect& rect1, const RotatedRect& rect2, OutputArray intersectingRegion)
  {
      CV_TRACE_FUNCTION();
      ... shift input ...
      std::vector <Point2f> intersection; intersection.reserve(24);
      int result = rotatedRectangleIntersection_(shiftedRect1, shiftedRect2, intersection);
      ... adjust 'intersection' points ...
      Mat(intersection).copyTo(intersectingRegion);  // ??? handle .empty() case
      return result;
  }
  ```

File: modules/imgproc/src/intersection.cpp [hunk 2]
```diff
@@ -300,7 +290,47 @@ int rotatedRectangleIntersection( const RotatedRect& rect1, const RotatedRect& r
...
+    RotatedRect shiftedRect1(rect1);
+    RotatedRect shiftedRect2(rect2);
+
+    // Move rectangles closer to origin
+    shiftedRect1.center -= averageCenter;
+    shiftedRect2.center -= averageCenter;
+
+    std::vector <Point2f> intersection; intersection.reserve(24);
+
+    const auto ret = _rotatedRectangleIntersection(shiftedRect1, shiftedRect2, intersection);
```
Comment (alalek) [inline modules/imgproc/src/intersection.cpp:320]: `auto` is C++11 feature.
3.4 branch requires C++98 code.

Use `int` here.

File: modules/imgproc/src/intersection.cpp [hunk 3]
```diff
@@ -300,7 +290,47 @@ int rotatedRectangleIntersection( const RotatedRect& rect1, const RotatedRect& r
...
+
+    // If return is not None, the intersection Points are shifted back to the original position
+    // and copied to the interesectingRegion
+    if (ret != INTERSECT_NONE)
+    {
+        for (auto &point : intersection)
```
Comment (alalek) [inline modules/imgproc/src/intersection.cpp:326]: `range for` is C++11 feature too.

```
for (size_t i = 0; i < intersection.size(); ++i)
{
    intersection[i] += averageCenter;
}
```

File: modules/imgproc/src/intersection.cpp [hunk 4]
```diff
@@ -300,7 +290,47 @@ int rotatedRectangleIntersection( const RotatedRect& rect1, const RotatedRect& r
...
+        {
+            point.x += averageCenter.x;
+            point.y += averageCenter.y;
+        }
+
+        Mat(intersection).copyTo(intersectingRegion);
+    }
```
Comment (alalek) [inline modules/imgproc/src/intersection.cpp:333]: Need to clean the output buffer if nothing is returned.

```
else
{
    intersectingRegion.release();
}
```

File: modules/core/src/types.cpp [hunk 5]
```diff
@@ -173,13 +173,13 @@ RotatedRect::RotatedRect(const Point2f& _point1, const Point2f& _point2, const P
 void RotatedRect::points(Point2f pt[]) const
 {
     double _angle = angle*CV_PI/180.;
-    float b = (float)cos(_angle)*0.5f;
-    float a = (float)sin(_angle)*0.5f;
+    double b = cos(_angle)*0.5;
+    double a = sin(_angle)*0.5;
```
Comment (alalek) [inline modules/core/src/types.cpp:177]: do we still need changes here?

**Review by alalek — COMMENTED** [review]
Thank you for update!

File: modules/core/src/types.cpp (hunk 5 above)
Comment (gasparitiago) [reply modules/core/src/types.cpp]: Actually, we don't need to change it, I'm reverting it.

File: modules/core/src/types.cpp [hunk 6]
```diff
@@ -173,13 +173,13 @@ RotatedRect::RotatedRect(const Point2f& _point1, const Point2f& _point2, const P
...
+    float a = sin(_angle)*0.5f;
 
-    pt[0].x = center.x - a*size.height - b*size.width;
-    pt[0].y = center.y + b*size.height - a*size.width;
-    pt[1].x = center.x + a*size.height - b*size.width;
-    pt[1].y = center.y - b*size.height - a*size.width;
+    pt[0].x = (float)(center.x - a*size.height - b*size.width);
+    pt[0].y = (float)(center.y + b*size.height - a*size.width);
+    pt[1].x = (float)(center.x + a*size.height - b*size.width);
+    pt[1].y = (float)(center.y - b*size.height - a*size.width);
```
Comment (alalek) [inline modules/core/src/types.cpp:182]: It makes sense to rollback all changes here.

**Review by alalek — COMMENTED** [review]
Looks good!
Please update PR's title and description before merge.

**Review by alalek — APPROVED** [review]
Thank you for contribution :+1:

**Pull request merged**